cmd/clusterctl/client/tree/discovery.go

```diff
@@ -151,21 +260,44 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 		}
 	}
 
-	// Handles orphan machines.
-	if len(machineMap) < len(machinesList.Items) {
-		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
-		tree.Add(workers, other)
+	return nil
+}
 
-		for i := range machinesList.Items {
-			m := &machinesList.Items[i]
-			if _, ok := machineMap[m.Name]; ok {
-				continue
-			}
-			addMachineFunc(other, m)
+func addMachinePoolsToObjectTree(cluster *clusterv1.Cluster, workers *unstructured.Unstructured, machinePoolList *expv1.MachinePoolList, tree *ObjectTree, options DiscoverOptions) {
+	for i := range machinePoolList.Items {
+		mp := &machinePoolList.Items[i]
+		addOpts := make([]AddObjectOption, 0)
+		if !options.ShowTemplates {
+			addOpts = append(addOpts, GroupingObject(true))
+		}
+		tree.Add(workers, mp, addOpts...)
+
+		if options.ShowTemplates {
+			bootstrapPlaceholder := constructPlaceholderObject(mp.Spec.Template.Spec.Bootstrap.ConfigRef)
+			tree.Add(mp, bootstrapPlaceholder, ObjectMetaName("BootstrapConfig"), NoEcho(true))
+
+			machinePlaceholder := constructPlaceholderObject(&mp.Spec.Template.Spec.InfrastructureRef)
+			tree.Add(mp, machinePlaceholder, ObjectMetaName("MachineInfrastructure"), NoEcho(true))
 		}
 	}
+}
 
-	return tree, nil
+func getResourceSetBindingInCluster(ctx context.Context, c client.Client, namespace string, name string) (*addonsv1.ClusterResourceSetBinding, error) {
+	if name == "" {
+		return nil, nil
+	}
```

Comment:
Is this required? same below

Reply:
Do you mean this function? I believe we need to get the binding to be able to find the associated resource sets.

Reply:
this comment is derived from the observation that this func is receiving 4 parameters but we are validating only one, which seems kind of odd. Also, what is the use case where the name can be nil?
TBH, being a private func this func I will drop name validation
same below

Reply:
I just referenced the existing functions like `getMachinesInCluster` which just have a name validation at the start.

```
func getMachinesInCluster(ctx context.Context, c client.Client, namespace, name string) (*clusterv1.MachineList, error) {
	if name == "" {
		return nil, nil
	}
        ...
}
```